mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -4356,6 +4907,542 @@ public static long pwritev (int fd, Iovec[] iov, long offset)
+		}
+		
+		public static int socketpair (AddrFamily domain, SockType type, SockProtocol protocol, out int sv0, out int sv1)
+		{
+			return socketpair (domain, type, 0, protocol, out sv0, out sv1);
+		}
+
+		public static int socketpair (AddrFamily domain, SockType type, SockFlags flags, SockProtocol protocol, int[] socket_vector)
+		{
+			if (socket_vector == null || socket_vector.Length != 2) {
+				SetLastError (Errno.EINVAL);
+				return -1;
+			}
+			int sv0, sv1;
+			int r = socketpair (domain, type, flags, protocol, out sv0, out sv1);
+			socket_vector[0] = sv0;
+			socket_vector[1] = sv1;
+			return r;
+		}
+
+		public static int socketpair (AddrFamily domain, SockType type, SockProtocol protocol, int[] socket_vector)
+		{
+			return socketpair (domain, type, 0, protocol, socket_vector);
+		}
+		
+		// sockatmark(2)
+		//    int sockatmark(int sockfd);
+		[DllImport (LIBC, SetLastError=true)]
+		public static extern int sockatmark (int socket);
+		
+		// listen(2)
+		//    int listen(int sockfd, int backlog);
+		[DllImport (LIBC, SetLastError=true)]
+		public static extern int listen (int socket, int backlog);
+		
+		// bind(2)
+		//    int bind(int sockfd, const struct sockaddr *addr, socklen_t addrlen);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_bind")]
+		public static unsafe extern int bind (int socket, byte* addr, long addrlen);
+
+		public static unsafe int bind (int socket, byte[] addr, long addrlen)
+		{
+			if (addrlen != 0 && addrlen > addr.Length)
+				throw new ArgumentOutOfRangeException ("addrlen", "addrlen > addr.Length");
+			fixed (byte* ptr = addr)
+				return bind (socket, ptr, addrlen);
+		}
+		
+		public static unsafe int bind (int socket, ISockaddr address)
+		{
+			var buffer = address.ToNativeArray ();
+			return bind (socket, buffer, buffer.Length);
+		}
+
+		// connect(2)
+		//    int connect(int sockfd, const struct sockaddr *addr, socklen_t addrlen);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_connect")]
+		public static unsafe extern int connect (int socket, byte* addr, long addrlen);
+
+		public static unsafe int connect (int socket, byte[] addr, long addrlen)
+		{
+			if (addrlen != 0 && addrlen > addr.Length)
+				throw new ArgumentOutOfRangeException ("addrlen", "addrlen > addr.Length");
+			fixed (byte* ptr = addr)
+				return connect (socket, ptr, addrlen);
+		}
+		
+		public static unsafe int connect (int socket, ISockaddr address)
+		{
+			var buffer = address.ToNativeArray ();
+			return connect (socket, buffer, buffer.Length);
+		}
+
+		// accept(2)
+		//    int accept(int sockfd, struct sockaddr *addr, socklen_t *addrlen);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_accept")]
+		public static unsafe extern int accept (int socket, byte* addr, ref long addrlen);
+
+		public static unsafe int accept (int socket, byte[] addr, ref long addrlen)
```

Comment:
I don't think so. The returned address basically has to be treated as a blob with unknown length because it might be a sockaddr_* type we don't know.